File: crates/domain/src/playback/resume_card.rs
use crate::catalog::{ArtworkRef, TitleId};

#[derive(Debug, Clone)]
pub struct ResumeCard {
    pub title: TitleId,
    pub display_title: String,
    pub artwork: Vec<ArtworkRef>,
    pub duration_ms: u64,
    pub progress_percent: u8,
    pub year: Option<u16>,
    pub series_title: Option<String>,
    pub series_artwork: Vec<ArtworkRef>,
    pub season_number: Option<u16>,
    pub episode_number: Option<u16>,
}

pub fn progress_percent(position_ms: u64, duration_ms: u64) -> u8 {
    (position_ms.saturating_mul(100) / duration_ms.max(1)).min(100) as u8
}

#[cfg(test)]
mod tests {
    use super::progress_percent;

    #[test]
    fn computes_bounded_percent() {
        assert_eq!(progress_percent(0, 1000), 0);
        assert_eq!(progress_percent(500, 1000), 50);
        assert_eq!(progress_percent(1000, 1000), 100);
        assert_eq!(progress_percent(2000, 1000), 100);
    }

    #[test]
    fn guards_zero_duration() {
        assert_eq!(progress_percent(500, 0), 100);
    }
}

File: clients/flutter/lib/model/common/resume_card.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'package:shadowmask/model/common/artwork.dart';
import 'package:shadowmask/model/common/title_ref.dart';

part 'resume_card.freezed.dart';
part 'resume_card.g.dart';

@freezed
abstract class ResumeCard with _$ResumeCard {
  const factory ResumeCard({
    required TitleRef title,
    required String displayTitle,
    Artwork? artwork,
    @Default(0) int durationMs,
    @Default(0) int progressPercent,
    int? year,
    String? seriesTitle,
    Artwork? seriesArtwork,
    int? seasonNumber,
    int? episodeNumber,
  }) = _ResumeCard;

  factory ResumeCard.fromJson(Map<String, dynamic> json) =>
      _$ResumeCardFromJson(json);
}
